### ISSUE TITLE
Invalid subquery with an incomplete GROUP BY clause accepted if part of LATERAL JOIN

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#24328]
### What version of Materialize are you using?

 v0.83.0-dev (224f7571e)

### What is the issue?

```
create table t1 (f1 integer);

# This fails as expected
select * from t1 JOIN (SELECT f1, count(*) from t1) as a1 USING (f1);
ERROR:  column "t1.f1" must appear in the GROUP BY clause or be used in an aggregate function

# This unexpectedly works
select * from t1 JOIN LATERAL (SELECT f1, count(*) from t1) as a1 USING (f1);
 f1 | count 
----+-------
(0 rows)
```

**Comment by philip-stoev** [comment]
See discussion in https://materializeinc.slack.com/archives/C02PPB50ZHS/p1704893756694129

**Comment by benesch** [comment]
Here's a one line repro that doesn't rely on lateral joins:

```
select (select (a, count(*)) from (select 1 as a) _) from (select 1 as a) _;
```

The problem seems to be, as @frankmcsherry explained to me earlier today, that in Materialize `a` binds successfully to the `a` in the outer scope, whereas PostgreSQL the `a` binds to the _inner_ a, which is an error to reference.

**Comment by benesch** [comment]
Fix is up in #24354.
